packages/run-protocol/src/feeDistributor.js

```diff
@@ -0,0 +1,302 @@
+      throw Error(`fee distributor timer died: ${done}`);
+    },
+  };
+
+  const collectionNotifier = E(timerService).makeNotifier(
+    0n,
+    collectionInterval,
+  );
+  observeNotifier(collectionNotifier, timeObserver).catch(e => {
+    console.error('fee distributor failed with', e);
+  });
+}
+
+/**
+ * @type {BuildFeeDistributor}
+ */
+export const buildDistributor = async (
+  collectors,
+  depositFacet,
+  timerService,
+  { epochInterval = 1n },
+) => {
+  startDistributing(
+    () =>
+      Promise.all(
+        collectors.map(collector =>
+          E(collector)
+            .collectFees()
+            .then(pmt => E(depositFacet).receive(pmt)),
+        ),
+      ),
+    timerService,
+    epochInterval,
+  );
+};
+
+/**
+ * @typedef {{ pushPayment: (payment: Payment, issuer: Issuer) => Promise<Amount>}} FeeDestination
+ *
+ * @param {Record<string, ERef<FeeDestination>>} [destinations]
+ * @param {Record<string, any>} [keywordShares]
+ */
+export const makeShareConfig = (destinations = {}, keywordShares = {}) => {
+  let totalShares = 0n;
+  const shares = Object.entries(destinations)
+    .filter(([_, dst]) => dst)
+    .map(([kw, destination]) => {
+      const share = keywordShares[kw];
+      assert.typeof(share, 'bigint', `${kw} must be a bigint; got ${share}`);
+      totalShares += share;
+      return {
+        share,
+        destination,
+      };
+    });
+  return { shares, totalShares };
+};
+
+/** @typedef {ReturnType<typeof makeShareConfig>} ShareConfig */
+
+/**
+ * shareFees tries to be tolerant of single points of failure.
+ *
+ * @param {Payment} payment
+ * @param {Issuer} issuer
+ * @param {ShareConfig} shareConfig
+ */
+export const shareFees = async (payment, issuer, { shares, totalShares }) => {
+  // Divide the payment into shares.
+  /** @type {Amount<'nat'>} */
+  const totalAmount = await E(issuer).getAmountOf(payment);
+  if (AmountMath.isEmpty(totalAmount)) {
+    return;
+  }
+  const { value: totalValue, brand } = totalAmount;
+
+  let remainder = totalValue;
+  /** @type {[ERef<FeeDestination>, Amount][]} */
+  const destinedAmountEntries = shares
+    .map(({ share, destination }, i) => {
+      const value =
+        i === shares.length - 1
+          ? remainder
+          : (totalValue * share) / totalShares;
+      remainder -= value;
+      /** @type {[ERef<FeeDestination>, Amount]} */
+      const entry = [destination, { value, brand }];
+      return entry;
+    })
+    .filter(([_, amt]) => !AmountMath.isEmpty(amt));
+
+  // Split the payment and assert we have a conservation of the total.
+  const sharedPayment = await E(issuer).splitMany(
+    payment,
+    destinedAmountEntries.map(([_dst, amt]) => amt),
+  );
+
+  // Send each nonempty payment to the corresponding destination.
+  await Promise.all(
+    destinedAmountEntries.map(([destination], i) =>
+      E(destination).pushPayment(sharedPayment[i], issuer),
+    ),
+  );
+};
+
+/** @type {ContractStartFn} */
```

Comment:
please type just the `zcf` param so the return type can be inferred. even better to define the terms like so:
[
```suggestion
/**
 * @param {ZCF} zcf
  */
```

even better to define the terms like so ](https://github.com/Agoric/agoric-sdk/blob/9156f614b7ea8ff8a1862194317d8ebf6324e9a2/packages/run-protocol/src/vaultFactory/liquidateIncrementally.js#L49-L62)

Reply:
Thanks for the advice.  I extracted the terms with `Parameters<...>`.